Modules/Sources/Yosemite/Tools/POS/POSCatalogPersistenceService.swift

```diff
@@ -68,6 +74,70 @@ final class POSCatalogPersistenceService: POSCatalogPersistenceServiceProtocol {
                       "\(variationImageCount) variation images, \(variationAttributeCount) variation attributes")
         }
     }
+
+    func persistIncrementalCatalogData(_ catalog: POSCatalog, siteID: Int64) async throws {
+        DDLogInfo("💾 Persisting incremental catalog with \(catalog.products.count) products and \(catalog.variations.count) variations")
+
+        try await grdbManager.databaseConnection.write { db in
+            let site = PersistedSite(id: siteID)
+            try site.insert(db, onConflict: .ignore)
+
+            for product in catalog.productsToPersist {
+                try product.insert(db, onConflict: .replace)
+            }
+
+            let productIDs = catalog.products.map { $0.productID }.map { String($0) }.joined(separator: ",")
+
+            try PersistedProductImage
+                .filter(sql: "\(PersistedProductImage.Columns.productID.name) IN (\(productIDs))")
+                .deleteAll(db)
+            for image in catalog.productImagesToPersist {
+                try image.insert(db, onConflict: .replace)
+            }
+
+            try PersistedProductAttribute
+                .filter(sql: "\(PersistedProductAttribute.Columns.productID.name) IN (\(productIDs))")
+                .deleteAll(db)
```

Comment:
For deletions, the alternative approach is to do `PersistedProductImage.filter { $0.productID == product.productID }.deleteAll(db)` in the product loop above.

I think when we move to compound keys (including the site ID) this might be an easier approach than constructing SQL strings for compound keys.

Reply:
Similar to above, I updated filtering by SQL with for loops comparing with ID in https://github.com/woocommerce/woocommerce-ios/pull/16106/commits/7dbd5b3a3c419d24253f2f5e5e082194c29a90b6.